src/macros.jl

```diff
@@ -237,7 +237,126 @@ macro grad(expr)
         end
     end |> esc
 end
+
+
+function funcall_fwd_to_rule(expr)
+    expr.head == :call || error("The rule should be in a format of a function call.")
+    func = expr.args[1]
+    has_tracked_data = false
+
+    args_l = Any[func]
+    args_r = Any[:(ReverseDiff.track)]
+    args_start = 2
+    if (expr.args[2].head == :parameters) # has kw args
+        push!(args_l, expr.args[2])
+        push!(args_r, expr.args[2])
+        args_start = 3
+    end
+    push!(args_r, func)
+    for arg in expr.args[args_start:end]
+        if arg.head == :(::)
+            arg_name = gensym(:arg)
+            arg_ex = :($arg_name::$(arg.args[end]))
+            push!(args_l, arg_ex)
+            push!(args_r, arg_ex)
+            if eval(arg.args[end]) <: TrackedReal || eval(arg.args[end]) <: TrackedArray
```

Comment:
Can we avoid `eval`?

Reply:
OK, I changed it to using an exhaustive list of tracked types to predicate the argument types now.